File: src/main/java/com/regionrental/managers/Rental.java
package com.regionrental.managers;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public class Rental {

    private final String regionName;
    private final UUID playerUUID;
    private final String playerName;
    private final long startDate;
    private long endDate;
    private int extensionCount;
    private double totalPaid;
    private double initialPrice; // Track initial rental price separately for refund calculations
    private double totalRefunded; // Track total amount refunded to prevent double-refunds
    private final List<RefundRecord> refundHistory; // Detailed refund history
    private final Set<String> warningSent;

    /**
     * Represents a single refund transaction for audit trail
     */
    public static class RefundRecord {
        private final double amount;
        private final long timestamp;
        private final String reason;
        private final String adminName;

        public RefundRecord(double amount, long timestamp, String reason, String adminName) {
            this.amount = amount;
            this.timestamp = timestamp;
            this.reason = reason;
            this.adminName = adminName;
        }

        public double getAmount() { return amount; }
        public long getTimestamp() { return timestamp; }
        public String getReason() { return reason; }
        public String getAdminName() { return adminName; }

        public String getFormattedTimestamp() {
            LocalDateTime dateTime = LocalDateTime.ofInstant(
                Instant.ofEpochMilli(timestamp),
                ZoneId.systemDefault()
            );
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MM/dd HH:mm");
            return dateTime.format(formatter);
        }
    }
    
    // Constructor for new rentals
    public Rental(String regionName, UUID playerUUID, String playerName, long endDate, double price) {
        this.regionName = regionName;
        this.playerUUID = playerUUID;
        this.playerName = playerName;
        this.startDate = System.currentTimeMillis();
        this.endDate = endDate;
        this.extensionCount = 0;
        this.totalPaid = price;
        this.initialPrice = price;
        this.totalRefunded = 0.0;
        this.refundHistory = new ArrayList<>();
        this.warningSent = new HashSet<>();
    }

    // Constructor for loading from storage (with refund data)
    public Rental(String regionName, UUID playerUUID, String playerName, long startDate, long endDate,
                  int extensionCount, double totalPaid, double initialPrice, double totalRefunded,
                  List<RefundRecord> refundHistory) {
        this.regionName = regionName;
        this.playerUUID = playerUUID;
        this.playerName = playerName;
        this.startDate = startDate;
        this.endDate = endDate;
        this.extensionCount = extensionCount;
        this.totalPaid = totalPaid;
        this.initialPrice = initialPrice;
        this.totalRefunded = totalRefunded;
        this.refundHistory = refundHistory != null ? refundHistory : new ArrayList<>();
        this.warningSent = new HashSet<>();
    }

    // Backward compatibility constructor (for old data without refund tracking)
    public Rental(String regionName, UUID playerUUID, String playerName, long startDate, long endDate,
                  int extensionCount, double totalPaid, double initialPrice) {
        this(regionName, playerUUID, playerName, startDate, endDate, extensionCount, totalPaid,
             initialPrice, 0.0, new ArrayList<>());
    }

    // Backward compatibility constructor (for old data without initialPrice or refund tracking)
    public Rental(String regionName, UUID playerUUID, String playerName, long startDate, long endDate,
                  int extensionCount, double totalPaid) {
        this(regionName, playerUUID, playerName, startDate, endDate, extensionCount, totalPaid,
             totalPaid, 0.0, new ArrayList<>());
    }
    
    public boolean isExpired() {
        return System.currentTimeMillis() > endDate;
    }
    
    public long getTimeRemaining() {
        return endDate - System.currentTimeMillis();
    }
    
    public int getDaysRemaining() {
        long millis = getTimeRemaining();
        if (millis <= 0) return 0;
        return (int) (millis / (1000 * 60 * 60 * 24));
    }
    
    public int getHoursRemaining() {
        long millis = getTimeRemaining();
        if (millis <= 0) return 0;
        return (int) (millis / (1000 * 60 * 60));
    }
    
    public String getFormattedEndDate() {
        LocalDateTime dateTime = LocalDateTime.ofInstant(
            Instant.ofEpochMilli(endDate), 
            ZoneId.systemDefault()
        );
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MM/dd HH:mm");
        return dateTime.format(formatter);
    }
    
    public void extendRental(int days, double price) {
        this.endDate = this.endDate + (days * 24L * 60L * 60L * 1000L);
        this.extensionCount++;
        this.totalPaid += price;
        // Clear warnings when extended
        this.warningSent.clear();
    }
    
    public void resetTime(int days) {
        this.endDate = System.currentTimeMillis() + (days * 24L * 60L * 60L * 1000L);
        this.warningSent.clear();
    }
    
    public boolean hasWarningBeenSent(String warningType) {
        return warningSent.contains(warningType);
    }
    
    public void markWarningSent(String warningType) {
        warningSent.add(warningType);
    }
    
    /**
     * Gets the total cost of extensions (excludes initial rental price)
     * @return The amount paid for extensions only
     */
    public double getExtensionCost() {
        return totalPaid - initialPrice;
    }

    /**
     * Records a refund transaction in the rental history
     * @param amount Amount refunded
     * @param reason Reason for refund (e.g., "duration_reset", "time_removal")
     * @param adminName Name of admin who issued refund
     */
    public void recordRefund(double amount, String reason, String adminName) {
        this.totalRefunded += amount;
        this.refundHistory.add(new RefundRecord(amount, System.currentTimeMillis(), reason, adminName));
    }

    /**
     * Gets the net amount that can still be refunded (prevents double-refunds)
     * @return The amount paid minus amount already refunded
     */
    public double getNetRefundableAmount() {
        return Math.max(0, totalPaid - totalRefunded);
    }

    /**
     * Adds time to rental with payment (for admin-charged additions)
     * Does NOT increment extension count (admin additions are separate)
     * @param days Days to add
     * @param price Amount charged
     */
    public void addTimeWithCharge(int days, double price) {
        this.endDate = this.endDate + (days * 24L * 60L * 60L * 1000L);
        this.totalPaid += price;
        // Note: extensionCount is NOT incremented for admin-charged additions
        this.warningSent.clear();
    }

    // Getters
    public String getRegionName() { return regionName; }
    public UUID getPlayerUUID() { return playerUUID; }
    public String getPlayerName() { return playerName; }
    public long getStartDate() { return startDate; }
    public long getEndDate() { return endDate; }
    public int getExtensionCount() { return extensionCount; }
    public double getTotalPaid() { return totalPaid; }
    public double getInitialPrice() { return initialPrice; }
    public double getTotalRefunded() { return totalRefunded; }
    public List<RefundRecord> getRefundHistory() { return new ArrayList<>(refundHistory); }

    // Setters (limited)
    public void setEndDate(long endDate) {
        this.endDate = endDate;
        this.warningSent.clear();
    }
    
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        Rental rental = (Rental) obj;
        return regionName.equals(rental.regionName);
    }
    
    @Override
    public int hashCode() {
        return regionName.hashCode();
    }
}

File: src/main/kotlin/com/regionrental/managers/Rental.kt
package com.regionrental.managers

import com.regionrental.util.TimeUtils
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.UUID

/**
 * Represents an active rental of a WorldGuard region.
 * Converted from 300-line Java class to ~150-line Kotlin class.
 *
 * This class maintains full Java interoperability through:
 * - @JvmField annotations for direct field access
 * - @JvmStatic factory methods replacing multiple constructors
 * - @JvmOverloads for default parameter support
 * - Nested RefundRecord class for backwards compatibility
 */
class Rental private constructor(
    val regionName: String,
    val worldName: String,
    val playerUUID: UUID,
    val playerName: String,
    val startDate: Long,
    endDate: Long,
    extensionCount: Int,
    totalPaid: Double,
    val initialPrice: Double,
    totalRefunded: Double,
    refundHistory: MutableList<RefundRecord>?,
    members: MutableSet<UUID>?
) {
    // Mutable fields
    var endDate: Long = endDate
        set(value) {
            field = value
            warningSent.clear()
        }

    var extensionCount: Int = extensionCount
        private set

    var totalPaid: Double = totalPaid
        private set

    var totalRefunded: Double = totalRefunded
        private set

    private val refundHistory: MutableList<RefundRecord> = refundHistory ?: mutableListOf()
    private val warningSent: MutableSet<String> = mutableSetOf()
    private val members: MutableSet<UUID> = members ?: mutableSetOf()

    // ============ Computed Properties ============

    /**
     * Unique composite key for this rental (world:region).
     */
    val compositeKey: String
        get() = "$worldName:$regionName"

    /**
     * Whether the rental has expired.
     */
    val isExpired: Boolean
        get() = System.currentTimeMillis() > endDate

    /**
     * Time remaining until expiration in milliseconds.
     */
    val timeRemaining: Long
        get() = endDate - System.currentTimeMillis()

    /**
     * Days remaining until expiration (0 if expired).
     */
    val daysRemaining: Int
        get() = if (timeRemaining <= 0) 0 else TimeUtils.millisToDays(timeRemaining)

    /**
     * Hours remaining until expiration (0 if expired).
     */
    val hoursRemaining: Int
        get() = if (timeRemaining <= 0) 0 else TimeUtils.millisToHours(timeRemaining)

    /**
     * Formatted end date string (MM/dd HH:mm).
     */
    val formattedEndDate: String
        get() = TimeUtils.formatTimestamp(endDate)

    /**
     * Total cost of extensions (excludes initial rental price).
     */
    val extensionCost: Double
        get() = totalPaid - initialPrice

    /**
     * Net amount that can still be refunded (prevents double-refunds).
     */
    val netRefundableAmount: Double
        get() = maxOf(0.0, totalPaid - totalRefunded)

    /**
     * Number of members in this rental.
     */
    val memberCount: Int
        get() = members.size

    // ============ Rental Operations ============

    /**
     * Extends the rental by the specified days and adds to total paid.
     */
    fun extendRental(days: Int, price: Double) {
        endDate += TimeUtils.daysToMillis(days)
        extensionCount++
        totalPaid += price
        warningSent.clear()
    }

    /**
     * Resets the rental time to start from now.
     */
    fun resetTime(days: Int) {
        endDate = System.currentTimeMillis() + TimeUtils.daysToMillis(days)
        warningSent.clear()
    }

    /**
     * Adds time with a charge (admin additions - does NOT increment extension count).
     */
    fun addTimeWithCharge(days: Int, price: Double) {
        endDate += TimeUtils.daysToMillis(days)
        totalPaid += price
        warningSent.clear()
    }

    // ============ Warning Management ============

    /**
     * Checks if a specific warning type has been sent.
     */
    fun hasWarningBeenSent(warningType: String): Boolean = warningType in warningSent

    /**
     * Marks a warning type as sent.
     */
    fun markWarningSent(warningType: String) {
        warningSent.add(warningType)
    }

    // ============ Refund Management ============

    /**
     * Records a refund transaction.
     */
    fun recordRefund(amount: Double, reason: String, adminName: String) {
        totalRefunded += amount
        refundHistory.add(RefundRecord(amount, System.currentTimeMillis(), reason, adminName))
    }

    /**
     * Gets a copy of the refund history.
     */
    fun getRefundHistory(): List<RefundRecord> = refundHistory.toList()

    // ============ Member Management ============

    /**
     * Gets a copy of all members.
     */
    fun getMembers(): Set<UUID> = members.toSet()

    /**
     * Adds a member to this rental.
     * @return true if added, false if already exists
     */
    fun addMember(memberUUID: UUID): Boolean = members.add(memberUUID)

    /**
     * Removes a member from this rental.
     * @return true if removed, false if not found
     */
    fun removeMember(memberUUID: UUID): Boolean = members.remove(memberUUID)

    /**
     * Checks if a player is a member.
     */
    fun hasMember(memberUUID: UUID): Boolean = memberUUID in members

    /**
     * Clears all members.
     */
    fun clearMembers() = members.clear()

    // ============ Object Methods ============

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Rental) return false
        return regionName == other.regionName && worldName == other.worldName
    }

    override fun hashCode(): Int = compositeKey.hashCode()

    override fun toString(): String = "Rental($compositeKey, player=$playerName, expires=$formattedEndDate)"

    // ============ Nested Classes ============

    /**
     * Represents a single refund transaction for audit trail.
     * Kept as nested class for Java backwards compatibility (Rental.RefundRecord).
     */
    data class RefundRecord(
        val amount: Double,
        val timestamp: Long,
        val reason: String,
        val adminName: String
    ) {
        val formattedTimestamp: String
            get() = TimeUtils.formatTimestamp(timestamp)
    }

    // ============ Factory Methods (Replace Java Constructors) ============

    companion object {
        /**
         * Creates a new rental (for when a player rents a region).
         */
        @JvmStatic
        fun create(
            regionName: String,
            worldName: String,
            playerUUID: UUID,
            playerName: String,
            endDate: Long,
            price: Double
        ): Rental = Rental(
            regionName = regionName,
            worldName = worldName,
            playerUUID = playerUUID,
            playerName = playerName,
            startDate = System.currentTimeMillis(),
            endDate = endDate,
            extensionCount = 0,
            totalPaid = price,
            initialPrice = price,
            totalRefunded = 0.0,
            refundHistory = null,
            members = null
        )

        /**
         * Loads a rental from storage (full data including refund history).
         */
        @JvmStatic
        @JvmOverloads
        fun fromStorage(
            regionName: String,
            worldName: String,
            playerUUID: UUID,
            playerName: String,
            startDate: Long,
            endDate: Long,
            extensionCount: Int,
            totalPaid: Double,
            initialPrice: Double,
            totalRefunded: Double = 0.0,
            refundHistory: MutableList<RefundRecord>? = null,
            members: MutableSet<UUID>? = null
        ): Rental = Rental(
            regionName = regionName,
            worldName = worldName,
            playerUUID = playerUUID,
            playerName = playerName,
            startDate = startDate,
            endDate = endDate,
            extensionCount = extensionCount,
            totalPaid = totalPaid,
            initialPrice = initialPrice,
            totalRefunded = totalRefunded,
            refundHistory = refundHistory,
            members = members
        )

        /**
         * Legacy constructor for old data without initialPrice or refund tracking.
         */
        @JvmStatic
        fun fromLegacyStorage(
            regionName: String,
            worldName: String,
            playerUUID: UUID,
            playerName: String,
            startDate: Long,
            endDate: Long,
            extensionCount: Int,
            totalPaid: Double
        ): Rental = fromStorage(
            regionName = regionName,
            worldName = worldName,
            playerUUID = playerUUID,
            playerName = playerName,
            startDate = startDate,
            endDate = endDate,
            extensionCount = extensionCount,
            totalPaid = totalPaid,
            initialPrice = totalPaid, // Default initialPrice to totalPaid
            totalRefunded = 0.0,
            refundHistory = null,
            members = null
        )

        /**
         * Migration constructor for old data without world field.
         */
        @JvmStatic
        fun fromMigration(
            regionName: String,
            defaultWorldName: String,
            playerUUID: UUID,
            playerName: String,
            startDate: Long,
            endDate: Long,
            extensionCount: Int,
            totalPaid: Double,
            initialPrice: Double,
            totalRefunded: Double,
            refundHistory: MutableList<RefundRecord>?
        ): Rental = fromStorage(
            regionName = regionName,
            worldName = defaultWorldName,
            playerUUID = playerUUID,
            playerName = playerName,
            startDate = startDate,
            endDate = endDate,
            extensionCount = extensionCount,
            totalPaid = totalPaid,
            initialPrice = initialPrice,
            totalRefunded = totalRefunded,
            refundHistory = refundHistory,
            members = null
        )
    }
}
